File: src/main/java/com/example/tasks/demo/builder/Cake.java
package com.example.tasks.demo.builder;

import lombok.Data;
import lombok.ToString;

@Data
@ToString
public class Cake {
    int age;
    String name;
    String color;

    private Cake(CakeBuilder cakeBuilder) {
        this.age = cakeBuilder.age;
        this.name = cakeBuilder.name;
        this.color = cakeBuilder.color;
    }

    public static class CakeBuilder {
        int age;
        String name;
        String color;

        CakeBuilder withAge(int age){
            this.age = age;
            return this;
        }

        CakeBuilder withName(String name){
            this.name = name;
            return this;
        }

        CakeBuilder withColor(String color){
            this.color = color;
            return this;
        }

        public Cake build(){
            return new Cake(this);
        }
    }
}

File: src/main/kotlin/builder/Cake.kt
package builder

class Cake private constructor(cakeBuilder: CakeBuilder) {
    private val age: Int = cakeBuilder.age
    private val name: String? = cakeBuilder.name
    private val color: String? = cakeBuilder.color

    override fun toString(): String {
        return "Cake(age=$age, name=$name, color=$color)"
    }

    data class CakeBuilder(
            var age: Int = 0,
            var name: String? = null,
            var color: String? = null) {

        fun withAge(age: Int) = apply { this.age = age }
        fun withName(name: String) = apply { this.name = name }
        fun withColor(color: String) = apply { this.color = color }
        fun build() =
            Cake(this)

    }

}